test/fail_compilation/impconv_array.d

```diff
@@ -0,0 +1,183 @@
+fail_compilation/impconv_array.d(164): Error: template instance `impconv_array.test1!(const(S3)[], immutable(S3)[], immutable(S3)[])` error instantiating
+fail_compilation/impconv_array.d(181):        instantiated from here: `test!(S3)`
+fail_compilation/impconv_array.d(153): Error: cannot implicitly convert expression `cast(const(S3)[])x ~ cast(const(S3)[])y` of type `const(S3)[]` to `S3[]`
+fail_compilation/impconv_array.d(164): Error: template instance `impconv_array.test1!(immutable(S3)[], S3[], S3[])` error instantiating
+fail_compilation/impconv_array.d(181):        instantiated from here: `test!(S3)`
+fail_compilation/impconv_array.d(153): Error: cannot implicitly convert expression `cast(const(S3)[])x ~ cast(const(S3)[])y` of type `const(S3)[]` to `immutable(S3)[]`
+fail_compilation/impconv_array.d(164): Error: template instance `impconv_array.test1!(immutable(S3)[], S3[], immutable(S3)[])` error instantiating
+fail_compilation/impconv_array.d(181):        instantiated from here: `test!(S3)`
+fail_compilation/impconv_array.d(153): Error: cannot implicitly convert expression `cast(const(S3)[])x ~ y` of type `const(S3)[]` to `S3[]`
+fail_compilation/impconv_array.d(164): Error: template instance `impconv_array.test1!(immutable(S3)[], const(S3)[], S3[])` error instantiating
+fail_compilation/impconv_array.d(181):        instantiated from here: `test!(S3)`
+fail_compilation/impconv_array.d(153): Error: cannot implicitly convert expression `cast(const(S3)[])x ~ y` of type `const(S3)[]` to `immutable(S3)[]`
+fail_compilation/impconv_array.d(164): Error: template instance `impconv_array.test1!(immutable(S3)[], const(S3)[], immutable(S3)[])` error instantiating
+fail_compilation/impconv_array.d(181):        instantiated from here: `test!(S3)`
+fail_compilation/impconv_array.d(182): Error: undefined identifier `_`
+  ---
+*/
+
+alias AliasSeq(TList...) = TList;
+
+@safe pure:
+
+void test1(X, Y, Z)()
+{
+    X x;
+    Y y;
+    Z z = x ~ y;
+}
+
+void test(T)()
+{
+    alias M = T[];
+    alias C = const(T)[];
+    alias I = immutable(T)[];
+    static foreach (X; AliasSeq!(M, C, I))
+        static foreach (Y; AliasSeq!(M, C, I))
+            static foreach (Z; AliasSeq!(M, C, I))
+                test1!(X,Y,Z)();
+}
+
+void test1654_ok()
+{
+    struct S1 { immutable(int)* x; }
+    test!(int);                 // ok, no indirections
+    test!(string);              // ok, immutable indirections
+    test!(S1);                  // ok, immutable indirections
+}
+
+void test1654_fail()
+{
+    struct S2 { int* x; }
+    struct S3 { const(int)* x; }
+    test!(char[]);              // fail, non-immutable indirections
+    test!(S2);                  // fail, non-immutable indirections
+    test!(S3);                  // fail, non-immutable indirections
+    _;                          // to always trigger an error during development
```

Comment:
Remove ?

Reply:
Will remove when https://github.com/dlang/dmd/pull/13040 is merged.

Reply:
Done.